tw-tasks-core/src/main/java/com/transferwise/tasks/health/ClusterWideTasksStateMonitor.java

```diff
@@ -84,33 +89,49 @@ public void init() {
                   taskHandleHolder.getValue().stop();
                   taskHandleHolder.getValue().waitUntilStopped(Duration.ofMinutes(1));
                 }
-                resetState();
+                resetState(false);
                 log.info("Monitoring of tasks state stopped.");
               });
         });
   }
 
-  protected void resetState() {
-    if (registeredMetricHandles != null) {
-      for (Object metricHandle : registeredMetricHandles) {
-        meterHelper.unregisterMetric(metricHandle);
+  protected void resetState(boolean forInit) {
+    stateLock.lock();
+    try {
+      if (registeredMetricHandles != null) {
+        for (Object metricHandle : registeredMetricHandles) {
+          meterHelper.unregisterMetric(metricHandle);
+        }
       }
-    }
 
-    stuckTasksCount = null;
+      stuckTasksCount = null;
 
-    erroneousTasksCount = null;
-    erroneousTasksCounts = new ConcurrentHashMap<>();
-    erroneousTasksCountPerType = new ArrayList<>();
-    registeredMetricHandles = new ArrayList<>();
-    taskInErrorStateHandles = new HashMap<>();
-    tasksHistoryLengthSeconds = new HashMap<>();
+      erroneousTasksCount = null;
+      erroneousTasksCounts = new ConcurrentHashMap<>();
```

Comment:
Since the read/write is guaraded by stateLock we can make it just a HashMap

Reply:
Fixed